### PR TITLE
Improvements to letsencrypt_mgmt_profile.py

### PR DESCRIPTION
**Pull request by patschi**
Notable changes:
* Added support for `Virtual Hosting` scenarios (SNI)
* Added possibility to update email contact
* Added more comments/setup instructions
* Added logging functionality to get at least some kind of output for debugging, if debug enabled
* Fixed some small things, like error messages, default values, typos, etc
* Fixed scenario where HTTP-to-HTTPS is enabled and token validation fails

Tested it with VH setup with SNI, RSA and ECDSA with a public-reachable test domain.

### CODE REVIEW

**Review by nikhilky — COMMENTED** [review]
Thanks for these enhancements. I had few comments

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 1]
```diff
@@ -1,19 +1,43 @@
...
 
-'''
 Parameters -
-    user            - Avi user name
-    password        - Password of the above user
-    tenant          - Avi tenant name
-    dryrun          - True/False. If True letsencrypt's staging server will be used.
+    user            - Avi user name (Default None)
+    password        - Password of the above user (Default None)
+    tenant          - Avi tenant name (default is the user's default tenant)
```
[the remark this answers is not in the document]
Comment (patschi) [reply cert_mgmt/letsencrypt_mgmt_profile.py]: You're right. Made the requested changes. Thanks!

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 2]
```diff
@@ -283,39 +358,84 @@ def _do_request_avi(url, method, data=None, error_msg="Error"):
...
+        self.terminal.write(message)
+        self.terminal.flush()
+        if self.debug:
+            self.log.write(message)
+            self.log.flush()
+
+    def setDebug(self, state):
+        self.debug = state
+        if state == True:
+            self.log = open("/tmp/letsencrypt-{}.log".format(datetime.datetime.now().strftime("%Y%m%d")), "a")
```
[the remark this answers is not in the document]
Comment (patschi) [reply cert_mgmt/letsencrypt_mgmt_profile.py]: As we discussed on Slack, I've now reverted the changes to `print()` statements, as it was previously.

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 3]
```diff
@@ -1,19 +1,43 @@
...
+    1. This content needs to be imported in the AVI Controller in the settings menu
+       at <<Templates - Security - Certificate Management>>.
+    2. Create at least following script params: user, password (sensitive).
+    3. Go to <<Templates - Security - SSL/TLS Certificates>>, click on <<Create>>
+       and then <<Application Certificate>>.
+    4. Specify a suitable name to identify this certificate in AVI Controller.
+       (Like sub.domain.tld RSA or sub.domain2.tld ECDSA)
+    5. Change <<Type>> to <<CSR>>
+    6. Set <<Common Name>> to the domain to which the certificate should be issued to.
+    7. Save and wait a few seconds for the certificate to be requested and imported.
```
Comment (chitr) [inline cert_mgmt/letsencrypt_mgmt_profile.py:18]: It should be mentioned that Certificate Management group from Step 1/2 should be add at Step 6.

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 4]
```diff
@@ -1,19 +1,43 @@
...
 Useful links -
     Ratelimiting - https://letsencrypt.org/docs/rate-limits/
+
+Source/Credits -
+    https://github.com/avinetworks/devops/blob/master/cert_mgmt/letsencrypt_mgmt_profile.py
+    Modified https://github.com/diafygi/acme-tiny/blob/master/acme_tiny.py (MIT license) for Avi Controller
 '''
 
-import os, subprocess, json, base64, binascii, time, hashlib, re
+import os, subprocess, json, base64, binascii, time, hashlib, re, ssl, sys, datetime, time
```
Comment (chitr) [inline cert_mgmt/letsencrypt_mgmt_profile.py:40]: Can you add imports as alphabetically  sorted ? or PEP8 stds.?

**Review by chitr — COMMENTED** [review]
Lets  log the Exceptions

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 5]
```diff
@@ -40,9 +64,15 @@ def _cmd(cmd_list, stdin=None, cmd_input=None, err_msg="Command Line Error"):
...
+    def _do_request(url, data=None, err_msg="Error", depth=0, verify=True):
         try:
-            resp = urlopen(Request(url, data=data, headers={"Content-Type": "application/jose+json", "User-Agent": "acme-tiny"}))
+            ctx = ssl.create_default_context()
+            if not verify:
+                # disable certificate verify
+                ctx.check_hostname = False
+                ctx.verify_mode = ssl.CERT_NONE
+            # open request.
+            resp = urlopen(Request(url, data=data, headers={"Content-Type": "application/jose+json", "User-Agent": "acme-tiny"}), context=ctx)
```
Comment (chitr) [inline cert_mgmt/letsencrypt_mgmt_profile.py:75]: Let's log all the Exceptions for following flow.

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 6]
```diff
@@ -172,44 +202,64 @@ def _do_request_avi(url, method, data=None, error_msg="Error"):
         token = re.sub(r"[^A-Za-z0-9_\-]", "_", challenge['token'])
         keyauthorization = "{0}.{1}".format(token, thumbprint)
 
-        # Update vs
+        # Get VSVIPs/VSs, based on FQDN
```
Comment (chitr) [inline cert_mgmt/letsencrypt_mgmt_profile.py:205]: Let's break this flow to smaller functions based on related functionality.

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 7]
```diff
@@ -290,32 +365,54 @@ def _do_request_avi(url, method, data=None, error_msg="Error"):
...
+
+    if contact != None and "@" in contact:
+        contact = [ "mailto:{}".format(contact) ] # contact must be array as of ACME RFC
+        print ("Contact set to: {}".format(contact))
+
     signed_crt = None
     try:
-        signed_crt = get_crt(user, password, tenant, api_version, csr_temp_file.name, directory_url=directory_url, contact=contact)
+        signed_crt = get_crt(user, password, tenant, api_version, csr_temp_file.name, disable_check=disable_check, directory_url=directory_url, contact=contact)
     finally:
```
Comment (chitr) [inline cert_mgmt/letsencrypt_mgmt_profile.py:414]: Should handle Exception here as well?

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 8]
```diff
@@ -225,31 +275,51 @@ def _do_request_avi(url, method, data=None, error_msg="Error"):
...
-            except (AssertionError, ValueError) as e:
-                raise ValueError("Wrote file to {0}, but couldn't download {1}: {2}".format('wellknown_path', 'wellknown_url', e))
+                try:
+                    maxVerifyAttempts = 5 # maximal amount of verification attempts
+                    # retrying logic. Otherwise race-condition can occurr between avi controller pushing config and the token validation request
+                    for verifyAttempt in range(maxVerifyAttempts):
+                        reqToken = _do_request(wellknown_url, verify=False)
+                        if reqToken[0] != keyauthorization:
+                            print ("Internal token validation failed, {0} of {1} attempts. Retrying in 2 seconds.".format((verifyAttempt + 1), maxVerifyAttempts))
+                            if maxVerifyAttempts == (verifyAttempt + 1): # attempts start at 0, therefore +1. If limit reached, raise exception.
```
Comment (chitr) [inline cert_mgmt/letsencrypt_mgmt_profile.py:313]: Lets reorganise the code by moving verifyAttempt outside loop.

File: cert_mgmt/letsencrypt_mgmt_profile.py [hunk 9]
```diff
@@ -172,44 +202,64 @@ def _do_request_avi(url, method, data=None, error_msg="Error"):
...
+
+        rsp = _do_request_avi("virtualservice/?search=({})".format(search_term), "GET").json()
         if rsp['count'] == 0:
             raise Exception("Could not find a VS with common name = {}".format(domain))
 
         vs_uuid = rsp["results"][0]["uuid"]
         print ("Found vs {} with fqdn {}".format(vs_uuid, domain))
-        # Check if the vs is servering on port 80
+
+        # Let's check if VS is enabled, otherwise challenge can never successfully complete.
```
Comment (chitr) [inline cert_mgmt/letsencrypt_mgmt_profile.py:224]: Lets move avi specific  flow to separate function.(for next iteration )

File: cert_mgmt/letsencrypt_mgmt_profile.py (hunk 7 above)
Comment (patschi) [reply cert_mgmt/letsencrypt_mgmt_profile.py:415]: When an exception inside `get_crt()` occurs, the script fails anyway, so not sure if handling exception at this point has advantages, or am I missing something?

File: cert_mgmt/letsencrypt_mgmt_profile.py (hunk 8 above)
Comment (patschi) [reply cert_mgmt/letsencrypt_mgmt_profile.py:314]: So you mean completely removing the loop and only using a `time.sleep` once?

**Comment by patschi** [comment]
I have had a chat with @nikhilky on this one. Main points:
1. Made requested changes in commits above
2. Breaking up the Avi specific flow to separate function is planned for a different PR

I have tested the current version, the last commit, working well on the latest 21.1.

There seems to be some issue when using multiple domains on one VS (basically using SANs) because of some restrictions of the `search` API functionality. I will check with Nikhil.

**Review by nikhilky — APPROVED** [review]
Discussed this PR with @chitr. The remaining items will be taken up in the next iteration.

**Pull request merged**